File: packages/python/tryaii_dre/centroids/generator.py
"""
Centroid generator -- creates benchmark centroids from training queries.

Centroids are the average embedding of all training queries for a benchmark.
They are used by the EmbeddingClassifier to measure how similar a user's
prompt is to each benchmark category.

Centroids are regenerated when the embedding model changes, because different
models produce different vector spaces.
"""

from __future__ import annotations

import json
import logging
import os
import tempfile
from pathlib import Path
from typing import Optional

import numpy as np

from tryaii_dre.embeddings.base import BaseEmbeddingProvider

logger = logging.getLogger("tryaii_dre.centroids")

# Bundled training queries
TRAINING_QUERIES_PATH = Path(__file__).parent / "data" / "training_queries.json"


class CentroidGenerator:
    """
    Generates and manages benchmark centroids.

    Centroids are the average embedding vector of representative queries
    for each benchmark. When a user sends a prompt, we compute cosine
    similarity between their prompt's embedding and each centroid to
    determine what kind of task they're asking about.

    Usage:
        generator = CentroidGenerator(embedding_provider)
        centroids = generator.generate()           # Generate from default queries
        generator.save(centroids, "path/to.json")  # Save to disk
        centroids = generator.load("path/to.json") # Load from disk
    """

    def __init__(self, embedding_provider: BaseEmbeddingProvider):
        self._provider = embedding_provider

    def generate(
        self,
        training_queries: Optional[dict[str, list[str]]] = None,
        show_progress: bool = True,
    ) -> dict[str, np.ndarray]:
        """
        Generate centroids from training queries.

        Args:
            training_queries: Dict of benchmark_name -> list of queries.
                              If None, uses bundled default queries.
            show_progress: Print progress during generation.

        Returns:
            Dict of benchmark_name -> centroid vector (numpy array).
        """
        if training_queries is None:
            training_queries = self._load_default_queries()

        centroids: dict[str, np.ndarray] = {}

        for benchmark, queries in training_queries.items():
            if show_progress:
                logger.info(f"Generating centroid for {benchmark} ({len(queries)} queries)...")

            # Embed all queries for this benchmark
            embeddings = self._provider.embed_batch(queries)

            # Centroid = average of all embeddings, then normalize
            centroid = np.mean(embeddings, axis=0)
            norm = np.linalg.norm(centroid)
            if norm > 0:
                centroid = centroid / norm

            centroids[benchmark] = centroid

        if show_progress:
            logger.info(f"Generated {len(centroids)} centroids "
                       f"(dim={self._provider.dimension})")

        return centroids

    def generate_from_custom(
        self,
        benchmark_name: str,
        queries: list[str],
    ) -> np.ndarray:
        """
        Generate a single centroid for a custom benchmark.

        Args:
            benchmark_name: Name of the benchmark.
            queries: Representative queries for this benchmark.

        Returns:
            Centroid vector (numpy array).
        """
        embeddings = self._provider.embed_batch(queries)
        centroid = np.mean(embeddings, axis=0)
        norm = np.linalg.norm(centroid)
        if norm > 0:
            centroid = centroid / norm
        return centroid

    def save(self, centroids: dict[str, np.ndarray], path: str | Path) -> None:
        """Save centroids to a JSON file."""
        path = Path(path)
        path.parent.mkdir(parents=True, exist_ok=True)

        data = {
            "metadata": {
                "model": self._provider.model_name,
                "dimension": self._provider.dimension,
                "benchmark_count": len(centroids),
            },
            "centroids": {
                name: vector.tolist() for name, vector in centroids.items()
            },
        }

        # Write to temp file then atomically rename
        fd, tmp_path = tempfile.mkstemp(dir=path.parent, suffix=".tmp")
        try:
            with os.fdopen(fd, "w") as f:
                json.dump(data, f)
            os.replace(tmp_path, path)
        except BaseException:
            os.unlink(tmp_path)
            raise

        logger.info(f"Saved {len(centroids)} centroids to {path}")

    @staticmethod
    def load(path: str | Path) -> tuple[dict[str, np.ndarray], dict]:
        """
        Load centroids from a JSON file.

        Returns:
            Tuple of (centroids dict, metadata dict).
        """
        with open(path, "r") as f:
            data = json.load(f)

        centroids = {
            name: np.array(vector, dtype=np.float32)
            for name, vector in data["centroids"].items()
        }

        return centroids, data.get("metadata", {})

    def _load_default_queries(self) -> dict[str, list[str]]:
        """Load bundled training queries."""
        with open(TRAINING_QUERIES_PATH, "r") as f:
            data = json.load(f)

        return {
            name: bench_data["queries"]
            for name, bench_data in data["benchmarks"].items()
        }

File: packages/node/src/centroids/generator.ts
/**
 * Centroid generator -- creates benchmark centroids from training queries.
 *
 * Centroids are the average embedding of all training queries for a benchmark.
 * They are used by the EmbeddingClassifier to measure how similar a user's
 * prompt is to each benchmark category.
 *
 * Centroids are regenerated when the embedding model changes, because different
 * models produce different vector spaces.
 */

import { readFileSync, writeFileSync, mkdirSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { BaseEmbeddingProvider } from '../embeddings/base.js';
import type { CentroidsJson, TrainingQueriesJson } from '../types.js';
import { vectorMean, vectorNormalize } from '../utils/math.js';

const currentDir = dirname(fileURLToPath(import.meta.url));

/** Path to bundled training queries. */
export const TRAINING_QUERIES_PATH = join(currentDir, 'data', 'trainingQueries.json');

/**
 * Generates and manages benchmark centroids.
 *
 * Centroids are the average embedding vector of representative queries
 * for each benchmark. When a user sends a prompt, we compute cosine
 * similarity between their prompt's embedding and each centroid to
 * determine what kind of task they're asking about.
 */
export class CentroidGenerator {
  private _provider: BaseEmbeddingProvider;

  constructor(embeddingProvider: BaseEmbeddingProvider) {
    this._provider = embeddingProvider;
  }

  /**
   * Generate centroids from training queries.
   *
   * @param trainingQueries - Dict of benchmark_name -> list of queries.
   *                          If undefined, uses bundled default queries.
   * @returns Dict of benchmark_name -> centroid vector (number array).
   */
  generate(trainingQueries?: Record<string, string[]>): Record<string, number[]> {
    if (!trainingQueries) {
      trainingQueries = this._loadDefaultQueries();
    }

    const centroids: Record<string, number[]> = {};

    for (const [benchmark, queries] of Object.entries(trainingQueries)) {
      // Embed all queries for this benchmark
      const embeddings = this._provider.embedBatch(queries);

      // Centroid = average of all embeddings, then normalize
      const centroid = vectorNormalize(vectorMean(embeddings));
      centroids[benchmark] = centroid;
    }

    return centroids;
  }

  /**
   * Async version of `generate` -- routes through the provider's async path
   * so it works with async-only providers like LocalEmbeddingProvider.
   * Sync providers work too via the base class's default async fallback.
   */
  async generateAsync(trainingQueries?: Record<string, string[]>): Promise<Record<string, number[]>> {
    if (!trainingQueries) {
      trainingQueries = this._loadDefaultQueries();
    }

    const centroids: Record<string, number[]> = {};

    for (const [benchmark, queries] of Object.entries(trainingQueries)) {
      const embeddings = await this._provider.embedBatchAsync(queries);
      const centroid = vectorNormalize(vectorMean(embeddings));
      centroids[benchmark] = centroid;
    }

    return centroids;
  }

  /**
   * Generate a single centroid for a custom benchmark.
   *
   * @param benchmarkName - Name of the benchmark.
   * @param queries - Representative queries for this benchmark.
   * @returns Centroid vector (number array).
   */
  generateFromCustom(benchmarkName: string, queries: string[]): number[] {
    const embeddings = this._provider.embedBatch(queries);
    return vectorNormalize(vectorMean(embeddings));
  }

  /**
   * Async version of `generateFromCustom`. Works with any provider
   * (sync via default fallback, async via its native async path).
   */
  async generateFromCustomAsync(benchmarkName: string, queries: string[]): Promise<number[]> {
    const embeddings = await this._provider.embedBatchAsync(queries);
    return vectorNormalize(vectorMean(embeddings));
  }

  /** Save centroids to a JSON file. */
  save(centroids: Record<string, number[]>, path: string): void {
    mkdirSync(dirname(path), { recursive: true });

    const data: CentroidsJson = {
      metadata: {
        model: this._provider.modelName,
        dimension: this._provider.dimension,
        benchmark_count: Object.keys(centroids).length,
      },
      centroids,
    };

    writeFileSync(path, JSON.stringify(data));
  }

  /** Load centroids from a JSON file. */
  static load(path: string): { centroids: Record<string, number[]>; metadata: CentroidsJson['metadata'] } {
    const raw = readFileSync(path, 'utf-8');
    const data: CentroidsJson = JSON.parse(raw);

    return {
      centroids: data.centroids,
      metadata: data.metadata,
    };
  }

  /** Load bundled training queries. */
  private _loadDefaultQueries(): Record<string, string[]> {
    const raw = readFileSync(TRAINING_QUERIES_PATH, 'utf-8');
    const data: TrainingQueriesJson = JSON.parse(raw);

    const queries: Record<string, string[]> = {};
    for (const [name, benchData] of Object.entries(data.benchmarks)) {
      queries[name] = benchData.queries;
    }
    return queries;
  }
}
